### PR TITLE
Adding fix for LD_LIBRARY_PATH in singularity mode

### PR DESCRIPTION
**Pull request by asp8200**
We had severe problems getting Nextflow to run the Sentieon-images from https://depot.galaxyproject.org/singularity/sentieon%3A202308--h43eeafb_0 
in singularity-mode.

Don Freed of Sentieon identified the problem, and Pontus found the fix presented in this PR.

I've tested the fix with Sentieon and I can now run the following test:
```
nextflow run main.nf -profile test_cache,software_license,<env> --sentieon_extension --aligner sentieon-bwamem --outdir results
```
with `<env>` set to `docker`, `singularity` and `conda`.

If the solution is satisfactory, then I'll implement it in the other Sentieon-modules.

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/sentieon/bwamem/main.nf [hunk 1]
```diff
@@ -5,7 +5,11 @@ process SENTIEON_BWAMEM {
 
     secret 'SENTIEON_LICENSE_BASE64'
 
-    container 'nf-core/sentieon:202308'
+    conda "bioconda::sentieon=202308"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/sentieon%3A202308--h43eeafb_0' :
```
Comment (maxulysse) [inline modules/nf-core/sentieon/bwamem/main.nf:10]:
```suggestion
        'https://depot.galaxyproject.org/singularity/sentieon:202308--h43eeafb_0' :
```

**Review by pontus — APPROVED** [review]
Looks ok.

File: modules/nf-core/sentieon/bwamem/main.nf [hunk 2]
```diff
@@ -47,6 +47,9 @@ process SENTIEON_BWAMEM {
         echo "Decoded and exported Sentieon test-license system environment variables"
     fi
 
+    LD_LIBRARY_PATH=/usr/local/lib/:\$LD_LIBRARY_PATH
```
Comment (pontus) [inline modules/nf-core/sentieon/bwamem/main.nf:50]: Maybe I should mention that if this is with conda, the path may be different. Not sure if the path works correctly with conda by default and that can be ignored.

File: modules/nf-core/sentieon/bwamem/main.nf [hunk 3]
```diff
@@ -47,6 +47,9 @@ process SENTIEON_BWAMEM {
...
+    export LD_LIBRARY_PATH
```
Comment (maxulysse) [inline modules/nf-core/sentieon/bwamem/main.nf:51]: Any risk into doing that in conda/docker.
or any risk of collision with local LD_LIBRARY_PATH?

File: modules/nf-core/sentieon/bwamem/main.nf (hunk 2 above)
Comment (maxulysse) [reply modules/nf-core/sentieon/bwamem/main.nf:50]: I think the default conda/docker works from what I can say.

Maybe let's try to do that only when using singularity/apptainer.

Comment (asp8200) [reply modules/nf-core/sentieon/bwamem/main.nf:50]: Ok, I'll try to do it that way. Testing right now.

File: modules/nf-core/sentieon/bwamem/main.nf [hunk 4]
```diff
@@ -30,6 +30,12 @@ process SENTIEON_BWAMEM {
     def sentieon_auth_mech_base64 = task.ext.sentieon_auth_mech_base64 ?: ''
     def sentieon_auth_data_base64 = task.ext.sentieon_auth_data_base64 ?: ''
 
+    if (workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container) {
```
Comment (pontus) [inline modules/nf-core/sentieon/bwamem/main.nf:33]: This is independent on how the container image is made available.

```suggestion
    if (workflow.containerEngine == 'singularity') {
```

**Review by maxulysse — APPROVED** [review]
LGTM

**Comment by asp8200** [comment]
CI-tests failing but I think that is all due to `API rate limit exceeded`.

File: modules/nf-core/sentieon/bwamem/main.nf [hunk 5]
```diff
@@ -5,7 +5,10 @@ process SENTIEON_BWAMEM {
 
     secret 'SENTIEON_LICENSE_BASE64'
 
-    container 'nf-core/sentieon:202308'
+    conda "bioconda::sentieon=202308"
```
Comment (maxulysse) [inline modules/nf-core/sentieon/bwamem/main.nf:8]:
```suggestion
    conda "${moduleDir}/environment.yml"
```
and add this to the `environment.yml` file
```
dependencies:
  - bioconda::sentieon=202308
```

File: modules/nf-core/sentieon/bwamem/main.nf [hunk 6]
```diff
@@ -21,15 +24,17 @@ process SENTIEON_BWAMEM {
...
     def prefix = task.ext.prefix ?: "${meta.id}"
     def sentieon_auth_mech_base64 = task.ext.sentieon_auth_mech_base64 ?: ''
     def sentieon_auth_data_base64 = task.ext.sentieon_auth_data_base64 ?: ''
 
+    if (workflow.containerEngine == 'singularity') {
+        fix_ld_library_path = 'LD_LIBRARY_PATH=/usr/local/lib/:\$LD_LIBRARY_PATH;export LD_LIBRARY_PATH'
+    } else {
+        fix_ld_library_path = ''
+    }
+
```
Comment (maxulysse) [inline modules/nf-core/sentieon/bwamem/main.nf:37]: I'll put that at the top of the script part, where we had the conda bit.
And can you add a comment about what is happening there.
And why

File: modules/nf-core/sentieon/bwamem/main.nf (hunk 5 above)
Comment (maxulysse) [reply modules/nf-core/sentieon/bwamem/main.nf]: we're missing the `environment.yml` file, no?

**Comment by DonFreed** [comment]
This seems OK to me.

I would prefer that this is fixed in the container (which seems to require a change in the biocontainers build process). However, I understand that might not be feasible. This seems like an acceptable workaround to get this working quickly.

**Pull request merged**